core/client/fs/src/main/java/alluxio/client/file/FileSystemContextReinitializer.java

```diff
@@ -89,51 +87,48 @@ public void onSuccess() {
   }
 
   /**
-   * Acquires the read lock to block reinitialization.
+   * Acquires the resource to block reinitialization.
    *
-   * When the context is being reinitialized, this call blocks until the reinitialization succeeds
-   * or fails. If it fails, an exception is thrown and the read lock will not be locked.
+   * When the context is being reinitialized, this call returns an empty optional.
+   * If the reinitialization fails, an exception is thrown.
    *
    * If there is an existing reinitialization exception, immediately throw an exception
    * without trying to block further reinitialization.
    *
    * @throws IOException when reinitialization fails before or during this method
-   * @return a resource holding the locked read lock
+   * @return the resource
    */
-  public LockResource acquireReadLockResource() throws IOException {
+  public Optional<CountResource> acquireBlockResource() throws IOException {
     Optional<IOException> exception = mExecutor.getException();
     if (exception.isPresent()) {
       throw exception.get();
     }
-    mLock.readLock().lock();
+    if (mCount.updateAndGet(c -> c >= 0 ? c + 1 : c) <= 0) {
+      return Optional.empty();
+    }
```

Comment:
Would it make sense to embed this logic and the allow logic to CountResource itself. CountResource by itself already is pretty much restricted to this scenario. So I feel it makes sense to make it expose its own allow/block functions and avoid exposing in its constructor a parameter for whether to increase/decrease the value.

Reply:
I imagine the `CountResource` can be used in other scenarios that need reference counting.